### DISCUSSION TITLE
Plotting results

### DISCUSSION THREAD

**Discussion by Lavanya-Bala** [Eomys/pyleecan#364] [Q&A]
Hi,
I am trying to plot the power output of my motor with the Electrical module and it runs into the following error

Traceback (most recent call last):
  File "", line 135, in <module>
    out_femm.plot_2D_Data("elec.Pem_av_ref", "time")
  File "\Methods\Output\Output\plot\plot_2D_Data.py", line 96, in plot_2D_Data

    fund_harm_dict = self.get_fund_harm(data)
  File "\Methods\Output\Output\getter\get_fund_harm.py", line 23, in get_fund_harm
    if data.symbol in ["B", "P", "Is", "Phi_{wind}"]:
AttributeError: 'numpy.float64' object has no attribute 'symbol'

Also, is it possible to plot the torque-speed/power_speed characteristics ?

**Comment by BonneelP** [comment]
Hello,

The issue is that elec.Pem_av_ref is defined as "Average Electromagnetic power" and is a float and not a Data object so you can't use this method. (computed here: https://github.com/Eomys/pyleecan/blob/master/pyleecan/Methods/Simulation/Electrical/comp_power.py)
By the way, we recently changed the way we handled plot in pyleecan cf #354 so now it would be output.elec.Is.plot_2D_Data("time") for instance (is you are using the latest version of pyleecan and SciDataTool). The tutorial were updated and the old syntax is still available. 

Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
Hello,
Do the updated tutorials have anything related to torque-speed characteristics plot ? just trying to figure out how to plot that performance curve

**Reply by BonneelP** [reply]
Hello,

Sadly we don't have yet dedicated torque-speed plots. But what you can do is use the VarLoadCurrent object to define an OP_matrix with 4 column (and is_torque=True) with each line corresponding to a speed, Id/Iq, torque combination (all the operating points are handled separately so nothing prevent to define a VarLoadCurrent on the torque/speed plane). At the end of the simulation you will have access (thought the Datakeeper) to the list of the results for each operating point and you can script a post-processing to gather the result in the Torque-speed plane. VarLoadCurrent will define Datakeeper for the "Tem_av_ref" for the "reference torque" from the OP_matrix and "Tem_av" for the computed average torque (all the default datakeeper are here: https://github.com/Eomys/pyleecan/blob/master/pyleecan/Methods/Simulation/VarSimu/gen_datakeeper_list.py). 

If you want to plot the torque as a function of speed for a "normal" VarLoadCurrent (with increasing speed) you can use xoutput.plot_multi("N0", "Tem_av")

Best regards,
Pierre

**Reply by Rutherfordio** [reply]
Hello,
First of all, congratulations on the software, really impressive.
Sorry to bother, but I'm also trying to do a torque-speed characteristic plot, however I'm being unsuccessful. Could you go a bit more into detail into how to set up the appropriate Id/Iq and torque combinations? 
Thank you very much,
Serge

**Reply by BonneelP** [reply]
Hello,

Thank you for your interest in pyleecan :)

In the above example we assume that you already have the torque speed plane that you want to explore. The torque in the OP_matrix is a "reference" and it can be different from the computed one by pyleecan. If you don't have the torque/speed data, you could create an OP_matrix with several N0/Id/Iq combination and at the end pyleecan will store all the torque for each combination in Xout.xoutput_dict["Tem_av"].result (you can take a look at Tests\Validation\Magnetics\test_torque.py (where we compare a "reference torque" to computed torque). Then you can reorganise the data to get your torque/speed plane. You can also take the opportunity to compute other Output by adding other Datakeeper (like force/flux harmonics) and it will be available in your computed torque/speed plane.

Hope it helps. Feel free to ask further questions otherwise ;)

Best regards,
Pierre

**Reply by Rutherfordio** [reply]
Hi,
I did as you said, and it worked, however I'm getting much higher torque values than my analytical calculations predict.
I suspect this is because the magnetic module is not limiting the voltage at all. 
How would I go about, limiting the Us voltage for all the simulations?

**Reply by BonneelP** [reply]
Hello,
If you already have the torque from another model, then you can define you OP matrix with (N0, Id, Iq, Tem_av_ref), each line is an Operating point (is_torque=True, type_OP_matrix = 1). Then pyleecan will also compute the torque and you will be able to compare the FEMM torque VS analytical torque as in Tests\Validation\Magnetics\test_torque.py. If the results are different then maybe you can check the machine definition (BH curbe, PM remanent flux density, number of turns, number of parallel circuits per phase, endwinding...). What kind of machine and analytical model are you using ? You can also check the flux computed by pyleecan with https://pyleecan.org/tuto_Plots.html (mmf plot is also available on the last step of the machine GUI).

By the way, pyleecan have 3 different output for torque:
- Tem_av | Average Electromagnetic torque [N.m]
- Tem_rip_norm | Peak to Peak Torque ripple normalized according to average torque (None if average torque=0) [-]
- Tem_rip_pp | Peak to Peak Torque ripple [N.m]
In the OP_matrix we define the "Reference average Electromagnetic torque [N.m]"

Best regards,
Pierre

**Reply by Rutherfordio** [reply]
Hello, how would I go about making it so that the VarLoadCurrent simulation also outputs the flux for each Id Iq combination?

**Reply by BonneelP** [reply]
Hello,
The "normal" behaviour of a multi-simulation is to run a reference simulation and then run the N generated simulations (Id/Iq from the OP_matrix). Then for each simulation we extract some results with the DataKeeper to avoid keeping everything. So to get the flux for all Id/Iq you have two options:
- A DataKeeper that store out.mag.B
- VarLoadCurrent.is_keep_all_output that will keep all the output of the N generated simulations and store it in xout.output_list. 

The second method is more efficient (you have the complete output object for each OP so you have access to all its post-processing methods) but can take lot's of memory. But if you are not storing the mesh for each Id/Iq it should be Ok.

Best regards,
Pierre
Ps: If the reference simulation matches one of the operating point from the OP_matrix pyleecan will skip the computation to avoid computing twice the same OP. So we have the habit to use:
simu.input.set_OP_from_array(simu.var_simu.OP_matrix, simu.var_simu.type_OP_matrix)
to enforce the first OP of the OP_matrix in the input and speed-up the simulation.

**Reply by Lavanya-Bala** [reply]
Hi, I am trying to find out if it is possible to change the scale of output plots instead of the automatic scaling that happens when plotting using plot_2D_data ?

**Reply by Lavanya-Bala** [reply]
Also, I wish to plot/print the torque harmonics ('Tem_h1' used in the optimisation process) but it is not a part of the OutMag class, could you please let me know how to go about this ?

**Reply by BonneelP** [reply]
Hello,

I'm not sure if I understood your first questions but plot_2D_data just opens a matplotlib figure so you can change the scale afterwards by interacting with the figure (fig = plt.gcf( )). You can also check plot_2D_data arguments here: https://github.com/Eomys/pyleecan/blob/master/pyleecan/Functions/Plot/plot_2D_Data.py

Regarding your second question, I guess that "Tem_h1" is a DataKeeper (or a OptiObjective which is the same thing), then you can find the corresponding values in Xoutput.xoutput_dict["Tem_h1"].result. Then let's say that your optimization has a design parameter named "H0" then you should be able to do plot_multi("H0", "Tem_h1") to have the curve Tem_h1 = fct(H0). More information on plot_multi available here: https://pyleecan.org/tuto_Operating_point.html#Iterating-on-several-Operating-Point

Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
Thanks for clarifying.
Sorry for the confusion, I understood your answer about the scale. 

Second, I perform optimisation with OptiObjective as maximising torque and minimising Tem_h1. So, after I obtain the optimal solutions, I wish to use the optimised values in my original motor design and visualise the improvement in results. In that case I am able to plot torque as usual with the magnetic module simulation. I was looking for how to do the same for Tem_h1 as it not part of the OutMag class.

**Reply by BonneelP** [reply]
Hello,

If you are looking for the performances and output of your best design, then you can set is_keep_all_output=True then in xouput.output_list[best_index] you will find the output object of the best simulation (by the way xouput.output_list[best_index].simu.save() will enable you to get the corresponding simulation and machine to run again for further postprocessing).
If I understood correctly, Tem_h1 is a scalar value corresponding to each simulation of the simulation so you can't "plot it" for your best simulation. The corresponding value will be available at Xoutput.xoutput_dict["Tem_h1"].result[best_index]. If Tem_h1 is an array then you will find the complete array at the same place (then you only need to find the corresponding x axis in the simulation).
On a side note Xoutput.mag.Tem contains the torque of the reference simulation (the simulation that you defined without any modification from the optimization) of your optimization and you can find the reference value for Tem_h1 at Xoutput.xoutput_dict["Tem_h1"].result_ref (to compare with the "best value").

Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
> Hello,
> 
> Thank you for your interest in pyleecan :)
> 
> In the above example we assume that you already have the torque speed plane that you want to explore. The torque in the OP_matrix is a "reference" and it can be different from the computed one by pyleecan. If you don't have the torque/speed data, you could create an OP_matrix with several N0/Id/Iq combination and at the end pyleecan will store all the torque for each combination in Xout.xoutput_dict["Tem_av"].result (you can take a look at Tests\Validation\Magnetics\test_torque.py (where we compare a "reference torque" to computed torque). Then you can reorganise the data to get your torque/speed plane. You can also take the opportunity to compute other Output by adding other Datakeeper (like force/flux harmonics) and it will be available in your computed torque/speed plane.
> 
> Hope it helps. Feel free to ask further questions otherwise ;)
> 
> Best regards,
> Pierre

Hello, 

Sorry for the trouble. Could you please clarify this further ? I don't have any reference values for Phi0_ref. Using the values in tutorial and plotting (plot_multi) Tem_av against N0 gives a curve which is similar to the (Tem_av vs Phi0) given in the tutorial. Whereas, what I am looking for is with an initial constant torque region which then diminishes to zero. Am I missing something here in setting up the matrix? I don't have analytical or reference values and I am trying to straightaway plot  the torque-speed behaviour of my motor.

**Reply by BonneelP** [reply]
Hello,

If you set :
`simu.var_simu.type_OP_matrix = 1`
Then you can set the OP_matrix with Id, Iq instead of I0, Phi0 (column order are then [N0, Id_ref, Iq_ref, Tem_av_ref])
You can set your torque curve (an initial constant torque region which then diminishes to zero) as the fourth column of this matrix and then set the corresponding N0, Id, Iq for each torque point on the other columns. 

If you don't already have the reference or analytical values, then we have everything to compute the torque according to a given N0, Id, Iq but we don't have a model to compute these value to achieve a particular torque. If you have any suggestion on which model we can use for that we can start investigating how to implement it. Which kind of machine are you working on ? 

Best regards,
Pierre

**Reply by Rutherfordio** [reply]
I ended up just using a basic MTPA setup for my capstone project, a basic motor design that aimed to compare Pyleecan simulations with analytical equations and other proprietary software. Although it still needs some refinement. 

The basic analytical equations for a finite drive like mine go like this:
![image](https://user-images.githubusercontent.com/7716357/124524499-91592380-ddfb-11eb-8a85-2699a29a4027.png)

Then you set up the correct zone changes to come up with the field weakening Id, Iq currents for a particular voltage limit.
![image](https://user-images.githubusercontent.com/7716357/124524562-d67d5580-ddfb-11eb-8468-fa6f0ac275a3.png)

Finally you use the Op matrix type one simulation:
![image](https://user-images.githubusercontent.com/7716357/124524750-8488ff80-ddfc-11eb-8569-115d55dc2e88.png)

And obtain a basic torque/speed characteristic. The strange rise in torque just after the nominal velocity can be explained by the fact that these currents are not tuned exactly for this motor. A couple of iterations, however, can resolve this problem, in this case taking the max torque point for the new constant torque period. 

![image](https://user-images.githubusercontent.com/7716357/124524676-425fbe00-ddfc-11eb-9690-435837efefcb.png)

**Reply by Lavanya-Bala** [reply]
> Hello,
> 
> If you set :
> `simu.var_simu.type_OP_matrix = 1`
> Then you can set the OP_matrix with Id, Iq instead of I0, Phi0 (column order are then [N0, Id_ref, Iq_ref, Tem_av_ref])
> You can set your torque curve (an initial constant torque region which then diminishes to zero) as the fourth column of this matrix and then set the corresponding N0, Id, Iq for each torque point on the other columns.
> 
> If you don't already have the reference or analytical values, then we have everything to compute the torque according to a given N0, Id, Iq but we don't have a model to compute these value to achieve a particular torque. If you have any suggestion on which model we can use for that we can start investigating how to implement it. Which kind of machine are you working on ?
> 
> Best regards,
> Pierre

I am working with an IPMSM machine

**Reply by Lavanya-Bala** [reply]
> I ended up just using a basic MTPA setup for my capstone project, a basic motor design that aimed to compare Pyleecan simulations with analytical equations and other proprietary software. Although it still needs some refinement.
> 
> The basic analytical equations for a finite drive like mine go like this:
> ![image](https://user-images.githubusercontent.com/7716357/124524499-91592380-ddfb-11eb-8a85-2699a29a4027.png)
> 
> Then you set up the correct zone changes to come up with the field weakening Id, Iq currents for a particular voltage limit.
> ![image](https://user-images.githubusercontent.com/7716357/124524562-d67d5580-ddfb-11eb-8468-fa6f0ac275a3.png)
> 
> Finally you use the Op matrix type one simulation:
> ![image](https://user-images.githubusercontent.com/7716357/124524750-8488ff80-ddfc-11eb-8569-115d55dc2e88.png)
> 
> And obtain a basic torque/speed characteristic. The strange rise in torque just after the nominal velocity can be explained by the fact that these currents are not tuned exactly for this motor. A couple of iterations, however, can resolve this problem, in this case taking the max torque point for the new constant torque period.
> 
> ![image](https://user-images.githubusercontent.com/7716357/124524676-425fbe00-ddfc-11eb-9690-435837efefcb.png)

Thank you, this was helpful

**Reply by Lavanya-Bala** [reply]
Hello,

I tried to do the same for Power speed curve using the OP_matrix for (N0,Id,Iq,T,P) and using 'is_power = false' to obtain the Power value vs increasing speed values. However, the simulation doesn't seem to run right as it outputs only one value of power at the end. Could you please let me know if I am missing something here or is there some other way to go about this ?

**Reply by BonneelP** [reply]
Hello,

Sorry for the long time to answer. In fact we have never used the "power column" of the OP matrix and the code wasn't available. I have added the corresponding code in PR #417 (I still needs to merge it and create a new release) and now it works if you set is_power=True and provide a 5 columns OP_matrix. However this set only the "reference power" that you input (same as for the reference torque). At the end it is available in xoutput.xoutput_dict["P_ref"].
I have added the method "OutMag.comp_power" method from my post bellow and the dedicated DataKeeper so the computed power is now also available in xoutput.xoutput_dict["P"].

Best regards,
Pierre

**Reply by faraz4714** [reply]
Dear all,
if anyone can answer me I will be grateful so much 
As anywhere was written about electromagnetic torque for IPMSM motor, we know it is equal to 
T = 3/2*N*(phim*Iq+(Ld - Lq)*Id*Iq).
and when I compute Ld and Lq and phim in pyleecan for a random operating point (in my case Id = -1 , Iq = 1 and rotor speed = 3000 and T_mag = 20 and N = 5) I got this value (Ld = 181*10^-6, Lq = 234*10^-6 and phim = 0.0045 ) and if anyone computes T, he/she gets T= 0.034
but if I compute the electromagnetic torque in pyleecan, I got 0.068 which is double 
and even if I change iq from 1 to 20 step by step and compute all the parameters, the torque is always double.

anyway, I simulated my motor in flux motor software for the different operating points and I got more or less the same results in pyleecan and the flux motor also my colleague measured the torque and it's the same.

but I am wondering and don't know why the formula is not working well because it's the general formula in the IPMSM motor.
Pierre, if you have any time could you please explain to me how the pyleecan can compute the torque in the magnetic model because I can understand the code 

Best regards,

Faraz Mirali

**Reply by SebGue** [reply]
Hello Faraz,

the formula you used to calculate the torque presumes that you use amplitued values for current and flux linkage. So I guess you calculated with RMS values. If you want to calculate with RMS values you got to remove the factor 1/2 from the formula. (The factor is derived from the amplitue to RMS converserion of current and flux linkage, i.e. 1/sqrt(2) * 1/sqrt(2).)

The torque in the MagFEMM module is calculate in FEMM internally by a weighted Maxwell Stress Tensor approch on the air gap of the machine. 

Hope this helps. Best regards,
Sebastian

**Reply by faraz4714** [reply]
hello Sebastian,
thank you for your answer you are right.
I check it and now everything is correct. can you please give me the reference for this formula? I should present it.
Best regards,
Faraz Mirali

**Reply by SebGue** [reply]
Hello Faraz,
sorry, I don't have any reference. But I guess you will find it in any book related to synchronous electrical machines.
BR, Sebastian

**Comment by Lavanya-Bala** [comment]
Hello,

I am trying to plot an efficiency curve with power input/output. Is there a way to do this with Pyleecan classes? The InputElec class doesn't seem to have any Input power parameters

**Reply by BonneelP** [reply]
Hello, 

For now there is nothing related to power (which is a shame). 

Regarding input power, it would be something like:
```
apparent power = qs*U_rms*conj(I_rms)
input power = 0.5*real(apparent power)
```
But we would need access to the voltage.

Regarding output power, you can use the following formula:
`2*pi*output.elec.N0/60*output.mag.Tem_av`
The best way would be to add the method OutMag.comp_power() (with the above formula) and then call this method in a DataKeeper. Do you think that you would be able to add this method ?

Best regards,
Pierre

**Reply by Lavanya-Bala** [reply]
Hello,

Thanks for the clarification, I will look into doing this. I also want to find the losses and I see the classes LossModelBertotti etc., Is there a tutorial or sample program on how these are implemented in a script used to simulate the motor ?

**Reply by RaphaelPile** [reply]
Not yet, but you can take a look at: https://github.com/Eomys/pyleecan/blob/master/Tests/Validation/Loss/test_FEMM_Loss.py

If you have some questions, maybe @SebGue can help you a little bit since he wrote the code.
